Add interface to DAQP

darnstrom opened the pull request:
This PR adds the dual active-set solver [DAQP](https://darnstrom.github.io/daqp/), which is aimed towards solving dense positive-definite QPs.

stephane-caron commented on qpsolvers/solvers/daqp_.py:
```diff
@@ -0,0 +1,182 @@
+    verbose: bool = False,
+    **kwargs,
+) -> Solution:
+    """Solve a quadratic program using daqp.
```
```suggestion
    """Solve a quadratic program using DAQP.
```

stephane-caron commented on qpsolvers/solvers/daqp_.py:
```diff
@@ -0,0 +1,182 @@
+
+"""Solver interface for `DAQP <https://github.com/darnstrom/daqp>`__.
+
+DAQP is a dual active-set algorithm implemented in C implementation. 
```
```suggestion
DAQP is a dual active-set algorithm implemented in C. 
```

stephane-caron commented on qpsolvers/solvers/daqp_.py:
```diff
@@ -0,0 +1,182 @@
+    sense = np.zeros(bupper.shape, dtype=c_int)
+    sense[ms+mineq:] = 5
+
+    x,obj,exitflag,info = daqp.solve(H,f,Atot,bupper,blower,sense)
```
This will not pass the CI linter check (should be `, ` rather than `,`). Can you run `black qpsolvers/` on the PR branch?

darnstrom replied:
When I ran black on the entire qpsolvers folder it also edited a lot of other files not related to DAQP, so I just ran it for the files that this PR has edited.

stephane-caron replied:
No worries, this looks good already. If there are some remaining linting errors I'll fix them when merging.

stephane-caron commented on qpsolvers/solvers/daqp_.py:
```diff
@@ -0,0 +1,182 @@
+        solution.z_box = info['lam'][:ms]  
+        solution.z = info['lam'][ms:ms+mineq] 
+        solution.y = info['lam'][ms+mineq:] 
+        #solution.extras = {
```
```suggestion
```

stephane-caron commented on qpsolvers/solvers/daqp_.py:
```diff
@@ -0,0 +1,182 @@
+        solution.z = info['lam'][ms:ms+mineq] 
+        solution.y = info['lam'][ms+mineq:] 
+        #solution.extras = {
+        #}
```
```suggestion
```

stephane-caron commented on qpsolvers/solvers/daqp_.py:
```diff
@@ -0,0 +1,182 @@
+    verbose: bool = False,
+    **kwargs,
+) -> Optional[np.ndarray]:
+    r"""Solve a quadratic program using daqp.
```
```suggestion
    r"""Solve a quadratic program using DAQP.
```

stephane-caron replied:
Or daqp, whichever is the preferred way to refer to the solver in writing.

stephane-caron commented on qpsolvers/solvers/daqp_.py:
```diff
@@ -0,0 +1,182 @@
+                & lb \leq x \leq ub
+        \end{array}\end{split}
+
+    It is solved using `daqp <https://pypi.python.org/pypi/daqp/>`__.
```
```suggestion
    It is solved using `DAQP <https://pypi.python.org/pypi/daqp/>`__.
```

stephane-caron reviewed (commented):
A new solver, great! :smiley: Thanks for opening this PR. It looks good to me, proposed changes are only cosmetic. Once it passes the CI it's good to go.

stephane-caron commented:
One last thing: since DAQP is on PyPI :+1: can you add it to:

- The ``testenv`` and ``testenv:coverage`` environments in ``tox.ini`` to run unit tests on it.
- The ``pip install`` line in the docs workflow in ``.github/workflows/docs.yml``.

Sorry this should have been documented already, I'll update the docs afterward in https://github.com/qpsolvers/qpsolvers/pull/200

stephane-caron commented:
By the way, if you add arm64 wheels to your PyPI distribution then feel free to add DAQP to the `dependencies` in `pyproject.toml``, so that it is installed by default with qpsolvers.

Without arm64 wheels you can add it to `[project.optional-dependencies]` in the same file. (arm64 wheels are needed for Raspberry Pi or Mac M* users.)

darnstrom commented:
The tox unit tests pass for me locally now! I also made the following additions:
* Since there are aarch64 wheels for DAQP now,  I added it to the dependencies in pyproject.toml.
* I added the corresponding article as a reference in the documentation
* Keyword arguments can now be used to change DAQP settings

stephane-caron commented:
Excellent work :+1: Thank you for contributing this PR. Let's release v3.3.0 to announce it.
